File: app/src/main/java/ru/melodin/fast/MainActivity.java
package ru.melodin.fast;

import android.Manifest;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import com.google.android.material.bottomnavigation.BottomNavigationView;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.ArrayList;
import java.util.List;

import ru.melodin.fast.api.UserConfig;
import ru.melodin.fast.api.VKApi;
import ru.melodin.fast.api.model.VKUser;
import ru.melodin.fast.common.AppGlobal;
import ru.melodin.fast.common.PermissionManager;
import ru.melodin.fast.common.ThemeManager;
import ru.melodin.fast.concurrent.AsyncCallback;
import ru.melodin.fast.concurrent.ThreadExecutor;
import ru.melodin.fast.current.BaseActivity;
import ru.melodin.fast.current.BaseFragment;
import ru.melodin.fast.database.CacheStorage;
import ru.melodin.fast.database.DatabaseHelper;
import ru.melodin.fast.fragment.FragmentConversations;
import ru.melodin.fast.fragment.FragmentFriends;
import ru.melodin.fast.fragment.FragmentItems;
import ru.melodin.fast.fragment.FragmentSettings;
import ru.melodin.fast.service.LongPollService;
import ru.melodin.fast.util.ArrayUtil;
import ru.melodin.fast.util.Util;
import ru.melodin.fast.util.ViewUtil;

public class MainActivity extends BaseActivity implements BottomNavigationView.OnNavigationItemSelectedListener, BottomNavigationView.OnNavigationItemReselectedListener {

    private FragmentConversations fragmentConversations = new FragmentConversations();
    private FragmentFriends fragmentFriends = new FragmentFriends();
    private FragmentItems fragmentItems = new FragmentItems();

    private int selectedId = -1;
    private BaseFragment selectedFragment;

    private BottomNavigationView navigationView;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        PermissionManager.setActivity(this);
        ViewUtil.applyWindowStyles(getWindow(), ThemeManager.getPrimaryDark());
        setTheme(ThemeManager.getCurrentTheme());
        VKApi.config = UserConfig.restore();
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        navigationView = findViewById(R.id.bottom_navigation_view);
        navigationView.setSelectedItemId(R.id.conversations);
        navigationView.setOnNavigationItemSelectedListener(this);
        navigationView.setOnNavigationItemReselectedListener(this);

        checkLogin();
        checkCrash();

        if (UserConfig.isLoggedIn()) {
            trackVisitor();
        }

        if (!PermissionManager.isGrantedPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
            PermissionManager.requestPermissions(44, Manifest.permission.WRITE_EXTERNAL_STORAGE);
        }

        EventBus.getDefault().register(this);
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putInt("selected_id", selectedId);
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        int selectedId = savedInstanceState.getInt("selected_id", -1);
        this.selectedId = selectedId;

        navigationView.setSelectedItemId(selectedId);
        selectedFragment = getFragmentById(selectedId);
    }

    private BaseFragment getFragmentById(int id) {
        switch (id) {
            case R.id.conversations:
                return fragmentConversations == null ? new FragmentConversations() : fragmentConversations;
            case R.id.items:
                return fragmentItems == null ? new FragmentItems() : fragmentItems;
            case R.id.friends:
                return fragmentFriends == null ? new FragmentFriends() : fragmentFriends;
            default:
                return null;
        }
    }

    private void loadUser() {
        ThreadExecutor.execute(new AsyncCallback(this) {
            VKUser user;

            @Override
            public void ready() throws Exception {
                user = VKApi.users().get().fields(VKUser.FIELDS_DEFAULT).execute(VKUser.class).get(0);
            }

            @Override
            public void done() {
                CacheStorage.insert(DatabaseHelper.USERS_TABLE, user);
                EventBus.getDefault().postSticky(new Object[]{"update_user", UserConfig.userId});
            }

            @Override
            public void error(Exception e) {
                Log.e("Error get user", Log.getStackTraceString(e));
            }
        });
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onReceive(Object[] data) {
        if (ArrayUtil.isEmpty(data)) return;

        String key = (String) data[0];
        if (key.equals(ThemeManager.KEY_THEME_UPDATE)) {
            applyStyles();
        }
    }

    private void startLoginActivity() {
        startActivity(new Intent(this, LoginActivity.class));
        finish();
    }

    private void checkLogin() {
        UserConfig.restore();
        if (!UserConfig.isLoggedIn()) {
            startLoginActivity();
        } else {
            loadUser();
            startService(new Intent(this, LongPollService.class));

            if (selectedFragment == null) {
                selectedFragment = fragmentConversations;
                selectedId = R.id.conversations;
            }

            replaceFragment(selectedFragment);
        }
    }

    private void trackVisitor() {
        if (!Util.hasConnection()) return;
        ThreadExecutor.execute(new AsyncCallback(this) {


            @Override
            public void ready() throws Exception {
                VKApi.stats().trackVisitor().execute();
            }

            @Override
            public void done() {

            }

            @Override
            public void error(Exception e) {
                Log.e("Error track visitor", Log.getStackTraceString(e));
            }
        });
    }

    private void checkCrash() {
        if (AppGlobal.preferences.getBoolean("isCrashed", false)) {
            final String trace = AppGlobal.preferences.getString("crashLog", "");
            AppGlobal.preferences.edit().putBoolean("isCrashed", false).putString("crashLog", "").apply();


            if (!AppGlobal.preferences.getBoolean(FragmentSettings.KEY_SHOW_ERROR, false))
                return;

            AlertDialog.Builder adb = new AlertDialog.Builder(this);
            adb.setTitle(R.string.warning);

            adb.setMessage(R.string.app_crashed);
            adb.setPositiveButton(android.R.string.ok, null);
            adb.setNeutralButton(R.string.show_error, (dialog, which) -> {
                AlertDialog.Builder adb1 = new AlertDialog.Builder(MainActivity.this);
                adb1.setTitle(R.string.error_log);
                adb1.setMessage(trace);
                adb1.setPositiveButton(android.R.string.ok, null);
                adb1.setNeutralButton(R.string.copy, (dialog1, which1) -> Util.copyText(trace));
                adb1.create().show();
            });

            adb.create().show();
        }
    }

    public void itemClick(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.groups:

                break;
            case R.id.menu:
                startActivity(new Intent(this, SettingsActivity.class));
                break;
            case R.id.left:
                showExitDialog();
                break;
        }
    }

    private void replaceFragment(Fragment fragment) {
        if (fragment == null) return;

        FragmentManager manager = getSupportFragmentManager();
        FragmentTransaction transaction = manager.beginTransaction();

        List<Fragment> fragments = manager.getFragments();
        List<String> classNames = new ArrayList<>(fragments.size());

        final int FRAGMENT_CONTAINER = R.id.fragment_container;

        if (ArrayUtil.isEmpty(fragments)) {
            transaction.add(FRAGMENT_CONTAINER, fragment, fragment.getClass().getSimpleName());
        } else {
            for (Fragment f : fragments) {
                transaction.hide(f);
                classNames.add(f.getClass().getSimpleName());
            }

            if (classNames.contains(fragment.getClass().getSimpleName())) {
                for (Fragment f : fragments)
                    if (f.getClass().getSimpleName().equals(fragment.getClass().getSimpleName())) {
                        transaction.show(f);
                        break;
                    }
            } else {
                transaction.add(FRAGMENT_CONTAINER, fragment, fragment.getClass().getSimpleName());
            }
        }
        transaction.commit();
    }

    private void showExitDialog() {
        AlertDialog.Builder adb = new AlertDialog.Builder(this);
        adb.setTitle(R.string.warning);
        adb.setMessage(R.string.exit_message);
        adb.setPositiveButton(R.string.yes, (dialog, which) -> {
            startLoginActivity();
            stopService(new Intent(this, LongPollService.class));
            UserConfig.clear();
            DatabaseHelper.getInstance().dropTables(AppGlobal.database);
            DatabaseHelper.getInstance().onCreate(AppGlobal.database);
        });
        adb.setNegativeButton(R.string.no, null);
        adb.create().show();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        selectedId = item.getItemId();
        switch (item.getItemId()) {
            case R.id.conversations:
                selectedFragment = fragmentConversations;
                break;
            case R.id.friends:
                selectedFragment = fragmentFriends;
                break;
            case R.id.menu:
                selectedFragment = fragmentItems;
                break;
        }

        replaceFragment(selectedFragment);
        return true;
    }

    @Override
    public void onNavigationItemReselected(@NonNull MenuItem item) {
        if (selectedFragment != null)
            selectedFragment.scrollToTop();
    }
}

File: app/src/main/java/ru/melodin/fast/MainActivity.kt
package ru.melodin.fast

import android.content.Intent
import android.os.Bundle
import android.util.SparseArray
import android.view.MenuItem
import android.view.View
import androidx.fragment.app.Fragment
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.microsoft.appcenter.AppCenter
import com.microsoft.appcenter.analytics.Analytics
import com.microsoft.appcenter.crashes.Crashes
import com.microsoft.appcenter.distribute.Distribute
import kotlinx.android.synthetic.main.activity_main.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import ru.melodin.fast.api.UserConfig
import ru.melodin.fast.api.VKApi
import ru.melodin.fast.common.*
import ru.melodin.fast.current.BaseActivity
import ru.melodin.fast.current.BaseFragment
import ru.melodin.fast.database.DatabaseHelper
import ru.melodin.fast.fragment.FragmentConversations
import ru.melodin.fast.fragment.FragmentItems
import ru.melodin.fast.fragment.FragmentMessages
import ru.melodin.fast.service.LongPollService
import ru.melodin.fast.util.ArrayUtil
import ru.melodin.fast.util.Keys
import ru.melodin.fast.util.Util
import java.util.*


open class MainActivity : BaseActivity(), BottomNavigationView.OnNavigationItemSelectedListener {

    private var fromRecreate: Boolean = false
    private var currentId = -1
    private var currentFragment: BaseFragment? = null

    private var stacks: SparseArray<Stack<Fragment>> = SparseArray()

    private lateinit var queue: ArrayList<String>

    var goBack = true

    companion object {
        val fragmentConversations = FragmentConversations()
        val fragmentItems = FragmentItems()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(ThemeManager.CURRENT_THEME)
        VKApi.config = UserConfig.restore()
        super.onCreate(savedInstanceState)
        initStacks()
        setContentView(R.layout.activity_main)

        navigationView.selectedItemId = R.id.navigation_conversations
        navigationView.setOnNavigationItemSelectedListener(this)

        checkLogin(savedInstanceState)
        CrashManager.checkCrash(this)

        if (UserConfig.isLoggedIn && savedInstanceState == null) {
            trackVisitor()
        }

        EventBus.getDefault().register(this)

        if (!BuildConfig.DEBUG) {
            AppCenter.start(
                application,
                "bd53321b-546a-4579-82fb-c68edb4feb20",
                Analytics::class.java
            )
        }

        AppCenter.start(
            application,
            "bd53321b-546a-4579-82fb-c68edb4feb20",
            Crashes::class.java
        )

        Distribute.setEnabledForDebuggableBuild(true)
        AppCenter.start(
            application,
            "bd53321b-546a-4579-82fb-c68edb4feb20",
            Distribute::class.java
        )
    }

    private fun initStacks() {
        queue = ArrayList()
        stacks.put(R.id.navigation_conversations, Stack())
        stacks.put(R.id.navigation_items, Stack())
    }

    private fun spliceStack(id: Int, size: Int) {
        stacks.get(id).setSize(size)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onReceive(data: Array<Any>) {
        if (ArrayUtil.isEmpty(data)) return

        when (data[0] as String) {
            Keys.THEME_UPDATE -> {
                fromRecreate = true
                applyStyles()
            }
            Keys.AUTHORIZATION_FAILED -> {
                val helper = DatabaseHelper.getInstance(this)
                val db = AppGlobal.database

                helper.dropTables(db)
                helper.onCreate(db)
                UserConfig.clear()

                stopService(Intent(this, LongPollService::class.java))
                finish()
                startActivity(Intent(this, LoginActivity::class.java))
            }
        }
    }

    private fun startLoginActivity() {
        startActivity(Intent(this, LoginActivity::class.java))
        finish()
    }

    private fun checkLogin(savedInstanceState: Bundle?) {
        UserConfig.restore()
        if (!UserConfig.isLoggedIn) {
            startLoginActivity()
        } else {
            if (savedInstanceState != null) return

            startLongPoll()

            currentFragment = fragmentConversations
            currentId = R.id.navigation_conversations

            replaceFragment(R.id.navigation_conversations, fragmentConversations, null)
        }
    }

    private fun startLongPoll() {
        startService(Intent(this, LongPollService::class.java))
    }

    private fun trackVisitor() {
        if (!Util.hasConnection() || BuildConfig.DEBUG) return
        TaskManager.execute {
            try {
                VKApi.stats().trackVisitor().execute()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (!fromRecreate) {
            stopService(Intent(this, LongPollService::class.java))
        }

        EventBus.getDefault().unregister(this)
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        if (item.itemId == currentId) {
            if (stacks.get(item.itemId).size > 1) {
                spliceStack(item.itemId, 1)
                showFragment(stacks.get(item.itemId).lastElement(), null)
            } else {
                val fragment = stacks.get(item.itemId).lastElement()
                if (fragment is BaseFragment)
                    fragment.scrollToTop()
            }
            return true
        }

        currentId = item.itemId

        if (stacks.get(item.itemId).size > 0) {
            queue.remove(item.itemId.toString())
            queue.add(item.itemId.toString())
            showFragment(stacks.get(item.itemId).lastElement(), null)
            return true
        }

        return when (item.itemId) {
            R.id.navigation_conversations -> {
                replaceFragment(item.itemId, fragmentConversations, null)
                true
            }
            R.id.navigation_items -> {
                replaceFragment(item.itemId, fragmentItems, null)
                true
            }
            else -> false
        }
    }

    fun replaceFragment(navId: Int, fragment: Fragment, arguments: Bundle?) {
        replaceFragment(navId, fragment, arguments, false)
    }

    fun replaceFragment(
        navId: Int,
        fragment: Fragment,
        arguments: Bundle?,
        showOnce: Boolean = false
    ) {
        var id = navId
        id = if (id == 0) currentId else id

        if (id != currentId) {
            navigationView.menu.findItem(id).isChecked = true
        }

        stacks.get(id).push(fragment)
        queue.remove(id.toString())
        queue.add(id.toString())

        arguments?.apply {
            putBoolean("top", stacks.get(id).size == 1)
        }

        showFragment(fragment, arguments, showOnce)
        checkBottomViewVisible(fragment)
    }

    private fun checkBottomViewVisible(fragment: Fragment) {
        if (fragment is BaseFragment) {
            if (fragment.isBottomViewVisible()) {
                showBottomView()
            } else {
                hideBottomView()
            }
        }
    }

    private fun showFragment(fragment: Fragment, arguments: Bundle?) {
        showFragment(fragment, arguments, false)
    }

    private fun showFragment(fragment: Fragment, arguments: Bundle?, showOnce: Boolean) {
        FragmentSelector.currentFragment = fragment
        FragmentSelector.selectFragment(supportFragmentManager, fragment, arguments, showOnce)
    }

    override fun onBackPressed() {
        if (currentFragment is FragmentMessages && !goBack) {
            currentFragment!!.onBackPressed()
        } else {
            stacks.get(currentId).pop()
            if (stacks.get(currentId).empty()) {
                queue.remove(currentId.toString())
                if (queue.size == 0) {
                    super.onBackPressed()
                    return
                }
                currentId = Integer.valueOf(queue[queue.size - 1])
            }

            val fragment = stacks.get(currentId).lastElement()

            showFragment(fragment, null)
            navigationView.menu.findItem(currentId).isChecked = true

            checkBottomViewVisible(fragment)
        }
    }

    private fun showBottomView() {
        navigationView.visibility = View.VISIBLE
    }

    private fun hideBottomView() {
        navigationView.visibility = View.GONE
    }
}